### PR TITLE
feat: auto provision dns and cloudfront

### PR DESCRIPTION
**Pull request by olizilla**
add cloudfront, acm, route53 magic to auto deploy domain and certs and cloudfront distribution per branch.

```sh
❯ curl https://pr12.autobahn.dag.haus
⁂ autobahn v0.0.0 (pr12)
- https://github.com/web3-storage/autobahn/commit/24d62c8ff159fca135a4513c822108b4fa2bc6bc
```

`stage` is set by sst via seed.run.
- It will be `prod` for production, where the domain will be `autobahn.dag.haus`
- It will be `staging` for staging, with the domain `staging.autobahn.dag.haus`
- it wll be `pr<num>` for PRs, with the domains `pr<num>.autobahn.dag.haus.

Notes
- I had to use a deprecated construct for creating the cert. The modern way doesn't let you specify which region to use, but cloudfront can **only** use certs made in `us-east-1`. The deprecated `DnsValidatedCertificate` construct lets us specify that.
- Properties of constructs like `fun.url` are wild magic strings at build time. You can't just pass it to `new URL` to extract the hostname, you have to ask cdk to do the string manipulation for you at deploy time using the cdk core fn utils.

Cloudfront distrbution for this PR: https://us-east-1.console.aws.amazon.com/cloudfront/v3/home?region=us-west-2#/distributions/E2LQZU00O4W4T8

### Refs

- https://github.com/aws/aws-cdk/tree/main/packages/aws-cdk-lib/aws-cloudfront#domain-names-and-certificates
- https://docs.aws.amazon.com/cdk/api/v1/docs/aws-cloudfront-readme.html
- https://paulallies.medium.com/deploy-your-static-react-app-to-aws-cloudfront-using-cdk-e53287e0052e
- https://bahr.dev/2020/09/01/multiple-frontends/

License: MIT

### CODE REVIEW

File: stacks/api.js [hunk 1]
```diff
@@ -23,7 +27,7 @@ export function API ({ stack }) {
 
   // Gotta us lambda fn url + cloudfront to get the streaming behavior we want
   // see: https://aws.amazon.com/blogs/compute/introducing-aws-lambda-response-streaming/
-  const fn = new Function(stack, 'fn', {
+  const fun = new Function(stack, 'fun', {
```
Comment (olizilla) [inline stacks/api.js:36]: never miss an opportunity for moar `fun`

**Review by alanshaw — APPROVED** [review]
👌 nice!

**Comment by olizilla** [comment]
This fails to deploy as the cloudfront distribution expects the cert to be provisioned in `us-east-1`

> Resource handler returned message: "Invalid request provided: AWS::CloudFront::Distribution: The specified SSL certificate doesn't exist, isn't in us-east-1 region, isn't valid, or doesn't include a valid certificate chain. 

Need to rethink this... sst + seed.run is focused on deploying to a single region. We need to deploy to `us-west-2`, but cloudfront can only create a distribution with an aws cert manager generated cert if that cert is created in `us-east-1`!

see: https://github.com/serverless-stack/sst/issues/1299

**Review by alanshaw — APPROVED** [review]
STILL GOOD

**Pull request merged**